lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,404 @@
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (buf_calc_free_space(b) < len) {
+		k_spin_unlock(&b->lock, key);
+		return -ENOMEM;
+	}
+
+	float *cur_head = b->append_head;
+	float *new_head = b->append_head + len;
+	int result = 0;
+	bool looped = false;
+
+	if (b->wait_data_size > 0) {
+		if (b->wait_data_size > len) {
+			b->wait_data_size -= len;
+		} else {
+			b->wait_data_size = 0;
+			result = PROCESS_BUF;
+		}
+	}
+
+	if (new_head >= buf_end) {
+		new_head -= ARRAY_SIZE(b->buf);
+		looped = true;
+	}
+
+	b->append_head = new_head;
+
+	k_spin_unlock(&b->lock, key);
+
+	if (looped) {
+		size_t buf_left = buf_end - cur_head;
+
+		memcpy(cur_head, data, buf_left * sizeof(float));
+		memcpy(buf_start, data + buf_left,
+		       (len - buf_left) * sizeof(float));
+	} else {
+		memcpy(cur_head, data, len * sizeof(float));
+	}
+
+	return result;
+}
+
+static void buf_get(struct data_buffer *b, float *b_res, size_t offset,
+		    size_t len)
+{
+	__ASSERT_NO_MSG((offset + len) <= INPUT_WINDOW_SIZE);
+
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	__ASSERT_NO_MSG(b->state == STATE_PROCESSING);
+
+	float *read_start = b->processing_head + offset;
+
+	k_spin_unlock(&b->lock, key);
+
+	float *read_end = read_start + len;
+	bool looped = false;
+
+	if ((read_end > buf_end) && (read_start < buf_end)) {
+		looped = true;
+	}
+
+	if (read_start >= buf_end) {
+		read_start -= ARRAY_SIZE(b->buf);
+	}
+
+	if (looped) {
+		size_t buf_left = buf_end - read_start;
+
+		memcpy(b_res, read_start, buf_left * sizeof(float));
+		memcpy(b_res + buf_left, buf_start,
+		       (len - buf_left) * sizeof(float));
+	} else {
+		memcpy(b_res, read_start, len * sizeof(float));
+	}
+}
+
+static int buf_processing_move(struct data_buffer *b, size_t move)
+{
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+	int res = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (b->state == STATE_READY) {
+		b->state = STATE_PROCESSING;
+	} else {
+		__ASSERT_NO_MSG(b->state != STATE_DISABLED);
+		k_spin_unlock(&b->lock, key);
+		return -EBUSY;
+	}
+
+	size_t max_move = buf_calc_max_processing_move(b);
+
+	b->processing_head += move;
+	if (b->processing_head >= buf_end) {
+		b->processing_head -= ARRAY_SIZE(b->buf);
+	}
+
+	size_t processing_end_move = move + INPUT_WINDOW_SIZE;
+
+	if (processing_end_move > max_move) {
+		b->wait_data_size = processing_end_move - max_move;
+	} else {
+		res = PROCESS_BUF;
+	}
+
+	k_spin_unlock(&b->lock, key);
+
+	return res;
+}
+
+size_t ei_ncs_get_frame_size(void)
+{
+	return INPUT_FRAME_SIZE;
+}
+
+size_t ei_ncs_get_window_size(void)
+{
+	return INPUT_WINDOW_SIZE;
+}
+
+int ei_ncs_add_data(const float *data, const size_t data_size)
+{
+	if (data_size % INPUT_FRAME_SIZE) {
+		return -EINVAL;
+	}
+
+	int res = buf_append(&ei_input, data, data_size);
+
+	if (res == PROCESS_BUF) {
```

Comment:
please do not mix two unrelated things - processing trigger indicator and error code

Reply:
Separated